### ISSUE TITLE
set terminal pngcairo enhanced font 'Verdana,8'

### ISSUE THREAD

**Issue by mopuru-murali** [juliojsb/sarviewer#4]
I'm getting gplot errors while generating graphs. I tried re-installeing gnuplot too after

    yum install cairo-devel pango-devel gd-devel

but no luck. I'm using centos 6.4 and gnuplot 4.2.6-2.el6.

	$> ./system_data_reader.sh

	List of sa* files available at this moment to retrieve data from:

	\-------------------------------------------
	File sa08 with data from Linux 2.6.32-358.23.2.el6.x86_64 (mmopuru-service.eng.vmware.com)      02/08/18        _x86_64_        (4 CPU)
	\-------------------------------------------

	Note that the number that follows the "sa" file specifies the day of the data collected by sar daemon
	Please select a sa* file from the listed above: sa08

	set terminal pngcairo enhanced font 'Verdana,8'
				 ^
	"loadaverage.gplot", line 15: unknown or ambiguous terminal type; type just 'set terminal' for a list

	set terminal pngcairo enhanced font 'Verdana,8'
				 ^
	"tasks.gplot", line 15: unknown or ambiguous terminal type; type just 'set terminal' for a list

	set terminal pngcairo enhanced font 'Verdana,8'
				 ^
	"cpu.gplot", line 16: unknown or ambiguous terminal type; type just 'set terminal' for a list

	set terminal pngcairo enhanced font 'Verdana,8'
				 ^
	"ram.gplot", line 16: unknown or ambiguous terminal type; type just 'set terminal' for a list

	set terminal pngcairo enhanced font 'Verdana,8'
				 ^
	"swap.gplot", line 16: unknown or ambiguous terminal type; type just 'set terminal' for a list

	set terminal pngcairo enhanced font 'Verdana,8'
				 ^
	"iotransfer.gplot", line 16: unknown or ambiguous terminal type; type just 'set terminal' for a list

	set terminal pngcairo enhanced font 'Verdana,8'
				 ^
	"proc.gplot", line 16: unknown or ambiguous terminal type; type just 'set terminal' for a list

	set terminal pngcairo enhanced font 'Verdana,8'
				 ^
	"contextsw.gplot", line 16: unknown or ambiguous terminal type; type just 'set terminal' for a list

	set terminal pngcairo enhanced font 'Verdana,8'
				 ^
	"netinterface.gplot", line 16: unknown or ambiguous terminal type; type just 'set terminal' for a list

	set terminal pngcairo enhanced font 'Verdana,8'
				 ^
	"sockets.gplot", line 16: unknown or ambiguous terminal type; type just 'set terminal' for a list

**Comment by mopuru-murali** [comment]
I tried installing gnuplot 4.4.2 from source too but no luck again throwing diff errros.

	--> sar -f /var/log/sa/sa09
	Linux 2.6.32-358.23.2.el6.x86_64 (mmopuru-service.eng.vmware.com)       02/09/2018      _x86_64_        (4 CPU)

	12:47:19 AM     CPU     %user     %nice   %system   %iowait    %steal     %idle
	12:56:17 AM     all      0.40      0.00      0.59      0.03      0.00     98.99
	01:00:01 AM     all      0.57      0.00      0.75      1.58      0.00     97.11
	Average:        all      0.45      0.00      0.63      0.48      0.00     98.43

	--> ./system_data_reader.sh
	List of sa* files available at this moment to retrieve data from:

	-------------------------------------------
	File sa08 with data from Linux 2.6.32-358.23.2.el6.x86_64 (mmopuru-service.eng.vmware.com)      02/08/18        _x86_64_        (4 CPU)
	File sa09 with data from Linux 2.6.32-358.23.2.el6.x86_64 (mmopuru-service.eng.vmware.com)      02/09/18        _x86_64_        (4 CPU)
	-------------------------------------------

	Note that the number that follows the "sa" file specifies the day of the data collected by sar daemon
	Please select a sa* file from the listed above: sa09

	plot '../../data/loadaverage.dat' using 1:2 title 'Tasks in run queue' with l ls 1,      ''                           using 1:3 title 'Total active tasks (processes + threads)' with l ls 2,      ''                           using 1:7 title 'Blocked tasks' with l ls 3
																																																																			  ^
	"tasks.gplot", line 42: warning: Skipping data file with no valid points

	plot '../../data/swap.dat' using 1:($2/1024) title 'Free' with l ls 1,      ''                    using 1:($3/1024) title 'Used' with l ls 2,
																																				  ^
	"swap.gplot", line 41: function to plot expected


	plot '../../data/iotransfer.dat' using 1:5 title 'Blocks read per second' with l ls 1,      ''                          using 1:6 title 'Blocks written per second' with l ls 2,
																																													 ^
	"iotransfer.gplot", line 41: function to plot expected


	plot '../../data/proc.dat' using 1:2 title 'Processes created per second' with l ls 1,
																						   ^
	"proc.gplot", line 39: function to plot expected


	plot '../../data/proc.dat' using 1:3 title 'Context switches performed per second' with l ls 1,
																									^
	"contextsw.gplot", line 40: function to plot expected


	plot '../../data/sockets.dat' using 1:($3+$7) title 'Total TCP sockets' with l ls 1,      ''                       using 1:3 title 'TCP sockets in use' with l ls 2,      ''                       using 1:4 title 'UDP sockets in use' with l ls 3,      ''                       using 1:7 title 'TCP sockets in TIME WAIT state' with l ls 4,
																																																																																					 ^
	"sockets.gplot", line 45: function to plot expected

**Comment by juliojsb** [comment]
Hello,

I haven't been able to reproduce the second issue.

Regarding the first issue, with **gd-devel** installed in a **CentOS 6** machine I can choose the terminal **png** to render PNG files:

```
gnuplot> set term
...
png  PNG images using libgd and TrueType fonts
...
```

With that in mind, in gnuplot scripts the following line:

`set terminal png enhanced font 'Verdana,8'`

Should be changed to:

`set terminal png`


I think the best option for rendering in this case would be Python (in **sarviewer.properties**, changing graph_generator="gnuplot" to graph_generator="matplotlib")

Regards,
Jota

**Comment by mopuru-murali** [comment]
'set term png' not working. Does 'pngcairo' help?

         png  PNG images using libgd and TrueType fonts
         pngcairo  png terminal based on cairo

**Comment by juliojsb** [comment]
**CentOS 6** repository ships with **gnuplot 4.2** and I don't see **pngcairo** terminal available. With **png** terminal the scripts work as expected without errors.

With **gnuplot 4.4.2** compiled pngcairo can be selected among the available terminals. I got the following warning:

```
[root@test-centos6 gnuplot]# gnuplot ram.gplot 

(process:6136): Pango-WARNING **: failed to choose a font, expect ugly output. engine-type='PangoRenderFc', script='common'

(process:6136): Pango-WARNING **: failed to choose a font, expect ugly output. engine-type='PangoRenderFc', script='latin'
```

I solved it by installing **dejavu-lgc-sans-fonts**

`yum install dejavu-lgc-sans-fonts`

### Error "function to plot expected"

Regarding that error, it happens because **gnuplot 4.4.2** seems to be stricter in the syntax of plot lines. Because of this, the following example would cause an error:

```
plot '../../data/sockets.dat' using 1:($3+$7) title 'Total TCP sockets' with l ls 1, \
     ''                       using 1:3 title 'TCP sockets in use' with l ls 2, \
     ''                       using 1:4 title 'UDP sockets in use' with l ls 3, \
     ''                       using 1:7 title 'TCP sockets in TIME WAIT state' with l ls 4, \
```

The last `, \` is interpreted by gnuplot as expecting a new line, which is wrong. To fix it:

```
plot '../../data/sockets.dat' using 1:($3+$7) title 'Total TCP sockets' with l ls 1, \
     ''                       using 1:3 title 'TCP sockets in use' with l ls 2, \
     ''                       using 1:4 title 'UDP sockets in use' with l ls 3, \
     ''                       using 1:7 title 'TCP sockets in TIME WAIT state' with l ls 4
```

I haven't found this syntax restrictions in more updated versions of gnuplot (like version 5.0). However, I have fixed it so it doesn't cause trouble in any version of gnuplot.

Clone again the repo and check, it should be working.

Regards,
Jota

**Comment by mopuru-murali** [comment]
It seems gnuplot 5+ required for sarviewer.
will you be updating 'installation' and 'requirements' sector in its documentation with necessary devels required?

**Comment by mopuru-murali** [comment]
Reading data files has problem when using gnuplot 4.4 in centos6

	 gnuplot44                                                         x86_64                                        4.4.2-2.el6                                               epel                                                        650 k
	 gnuplot44-common                                                  x86_64                                        4.4.2-2.el6                                               epel                                                        571 k

	--> ./data_collector.sh
	Please specify the number of samples to take-> 10
	Please specify the sample interval (take sample every X seconds)-> 2
	Taking 10 samples with 2 seconds interval
	Total time to collect all data: 20 seconds
	----------------------------------
	>>> Collecting data
	>>> Please wait until data collection is completed
	----------------------------------
	- You can abort this script with Ctrl+C, but have in mind the data will stop being collected when you cancel it.
	- You will also need to manually launch script plotter.sh to generate the graphs.
	----------------------------------
	>>> Just 5 seconds more while processing all data
	----------------------------------
	Done!

	plot '../../data/loadaverage.dat' using 1:2 title 'Tasks in run queue' with l ls 1,      ''                           using 1:3 title 'Total active tasks (processes + threads)' with l ls 2,      ''                           using 1:7 title 'Blocked tasks' with l ls 3
																																																																			  ^
	"tasks.gplot", line 42: warning: Skipping data file with no valid points

	plot '../../data/netinterface.dat' using 1:5 title 'Kilobytes received per second' with l ls 1,      ''                            using 1:6 title 'Kilobytes sent per second' with l ls 2
																								 ^
	"netinterface.gplot", line 41: warning: Skipping data file with no valid points

	plot '../../data/netinterface.dat' using 1:5 title 'Kilobytes received per second' with l ls 1,      ''                            using 1:6 title 'Kilobytes sent per second' with l ls 2
																																															 ^
	"netinterface.gplot", line 41: warning: Skipping data file with no valid points

	plot '../../data/netinterface.dat' using 1:5 title 'Kilobytes received per second' with l ls 1,      ''                            using 1:6 title 'Kilobytes sent per second' with l ls 2
																																															  ^
	"netinterface.gplot", line 41: x range is invalid

**Comment by juliojsb** [comment]
Gnuplot 5+ is not needed for this to work. Terminal pngcairo is available since gnuplot 4.4 as stated in the oficial documentation:

> Two new drivers based on the cairo and pango libraries are included, pngcairo and pdfcairo.  These are alternatives to the older libgd-based png driver and the older PDFLib-based pdf driver.  The figures in the pdf version of this manual were prepared using the pdfcairo terminal driver.

http://www.gnuplot.info/docs_4.4/gnuplot.pdf

Documentation updated with gnuplot 4.4+ requirement.

I think the last error reported is not related to gnuplot 4.4. I have checked and it seems it is because of the way sysstat is generating data. For example:

```
[root@test-centos6 sarviewer]# sar -V
sysstat version 9.0.4
(C) Sebastien Godard (sysstat <at> orange.fr)

[root@test-centos6 sarviewer]# sar -q 1 1
Linux 2.6.32-696.20.1.el6.x86_64 (test-centos6) 	02/28/2018 	_x86_64_	(1 CPU)

05:29:36 PM   runq-sz  plist-sz   ldavg-1   ldavg-5  ldavg-15
05:29:37 PM         0        83      0.00      0.00      0.00
Average:            0        83      0.00      0.00      0.00
```

The gnuplot tasks script looks for one more column that doesn't exist. It seems at least **sysstat 10.1.5+** is required. Compiling it seems to solve the issue:

```
[root@test-centos6 sarviewer]# sar -V
sysstat version 10.1.5
(C) Sebastien Godard (sysstat <at> orange.fr)

[root@test-centos6 sarviewer]# sar -q 1 1
Linux 2.6.32-696.20.1.el6.x86_64 (test-centos6) 	02/28/18 	_x86_64_	(1 CPU)

17:48:32      runq-sz  plist-sz   ldavg-1   ldavg-5  ldavg-15   blocked
17:48:33            0        82      0.00      0.03      0.02         0
Average:            0        82      0.00      0.03      0.02         0

[root@test-centos6 sarviewer]# ./data_collector.sh 
Please specify the number of samples to take-> 10
Please specify the sample interval (take sample every X seconds)-> 2
Taking 10 samples with 2 seconds interval
Total time to collect all data: 20 seconds
----------------------------------
>>> Collecting data
>>> Please wait until data collection is completed
----------------------------------
- You can abort this script with Ctrl+C, but have in mind the data will stop being collected when you cancel it.
- You will also need to manually launch script plotter.sh to generate the graphs.
----------------------------------
>>> Just 5 seconds more while processing all data
----------------------------------
Done!
```

Regards,
Jota
